Update missing items in context.json

davidlj95 opened the pull request:
After processing several assertions using the 2.0 and 2.1 context.json, I've found two missing items in the context that are accepted by the Blockcerts and Chainpoint standards:

- [`ETHData` type (possible value of `.signature.anchor.type`)](https://chainpoint.org/)
- [`chain` field (optional field in `.signature.anchor`)](https://github.com/blockchain-certificates/cert-schema/blob/master/docs/merkleProofSignatureExtension_schema.md#chain-string)

downie commented on scripts/convert-certificate.py:
```diff
@@ -0,0 +1,47 @@
+        data['assertion']['type'] = 'Assertion'
+
+        # fix images
+        if 'image' in data['certificate']:
```
This is such a clear example of the difference between the formats. Thanks for writing this.

kayaelle commented:
Took a look at examples/1.1.0/sample_unsigned_cert-1.1.0.json in terms of Open Badges compliance. Some initial thoughts:

recipient: 

- familyName, givenName, & pubkey aren’t recipient properties. Can see why you need pubkey of course.  familyName & givenName (or something similar) is being proposed for 2.0
- why is recipient a separate object vs a required property of assertion?

certificate: 

- why not call it badgeClass?
- missing name (required)
- missing criteria (required)
- title & subtitle aren’t badge class properties

assertion:

- can see why verify isn’t part of the assertion but maybe there is a better way to go about this. Nate may have some additional ideas. For instance, perhaps there is a verify object and then use another name for what you currently have as the verify object that does the wrapping? Need to think on this some.
- How about considering nesting certificate (or badge class) within assertion much like issuer is nested inside certificate?


Images (assertion and certificate): can you explain the intention of these property values?

One thought is to utilize extensions to make this compliant: https://openbadgespec.org/extensions/

kimdhamilton commented:
@kayaelle thanks for the feedback! 

First some general explanation:
The 1.1 schema is what we've already released -- we have a couple of deployments that we need to support. A lot of this exercise was trying to understand how our current schema differed from OBI, and how to get it to converge to OBI 1.1 in our subsequent releases.  Once we understood the structural problems with getting our schema to match OBI 1.1, we decided to:

**1.2 release:**
a. fix leaf properties where our types clash with OBI's (details below)
b. minimize breaking changes compared to our 1.1 schema until we've figured out how to address the structural problems (https://github.com/digital-certificates/cert-schema/issues/4)
c. rework our JSON schema plus introduce a JSON LD schema, both of which are now more clearly anchored to OBI corresponding OBI schemas (i.e. at least we know where we conflict now)

Our 1.2 is a effectively lot of schema housekeeping so we can start enforcing compliance against our schema. More importantly, we have a clear understanding and documentation of our OBI non-compliance issues. This allows us to work with Nate, etc on the structural changes to get our schema to comply with OBI in a subsequent version (our version 1.3).

Because of these improvements, we want to encourage people to upgrade to our 1.2 schema (from 1.1) and minimize breaking changes in between versions.

Specific responses:

>>_familyName, givenName, & pubkey aren’t recipient properties. Can see why you need pubkey of course. familyName & givenName (or something similar) is being proposed for 2.0_

We originally had familyName and givenName from https://schema.org/Person. I'd like to address this in schema 1.3 to avoid breaking changes.

>>_why is recipient a separate object vs a required property of assertion?_

I'd like to address this along with 1.3 (breaking change)

>>_certificate:_
>>_why not call it badgeClass?_

we'll address this in 1.3  (breaking change)

>>_missing name (required)_

I realized that our 1.1 version was called title instead of name. They are functionally the same. We'll address this in 1.3

>>_missing criteria (required)_

I added it to our 1.2 schema, but it's still optional. We'll address making it required with 1.3

>>_title & subtitle aren’t badge class properties_

title is related to name; subtitle is an extension. I made subtitle optional in 1.2. 

>> _assertion:_

>>_can see why verify isn’t part of the assertion but maybe there is a better way to go about this. Nate may have some additional ideas. For instance, perhaps there is a verify object and then use another name for what you currently have as the verify object that does the wrapping? Need to think on this some._
>>_How about considering nesting certificate (or badge class) within assertion much like issuer is nested inside certificate?_

great suggestions, and we're open to any restructuring that you guys are happy with. We want to start thinking about 1.3 asap.

>>_Images (assertion and certificate): can you explain the intention of these property values?_

Yes, these are interesting. Ours are base64-encoded images and not urls. This was originally done to avoid as many external links as possible. Because we hash the certificate, we prefer to inline as many elements as possible (as opposed to link). The main concern we had is the "permanent" hosting burden for each URL in the cert; otherwise the document becomes less useful. 

>>_One thought is to utilize extensions to make this compliant: https://openbadgespec.org/extensions/_

@ottonomy said he had some ideas about this, and was going to follow up soon

davidlj95 committed:
update missing items in context.json

add ETHData type (possible value of .signature.anchor.type)
add chain field (optional field in .signature.anchor)

lemoustachiste commented:
This is pretty old, but, have you ever experienced these properties to come as "unmapped"?

davidlj95 commented:
Oops, yeah that was a long time ago . Brings good memories though :) 

I recall some assertions failing to verify when parsing the JSON LD because of those missing items

Recall using these sometimes as example:

https://github.com/AnthonyRonning/https-github.com-labnol-files/blob/master/Ethereum-Main.json

https://github.com/blockchain-certificates/cert-schema/blob/f348277ec18330a46f6a8a9566ce1cbcd4d56b12/examples/2.1/sample_valid.json

And these to parse the assertion or verifying the certificate
https://github.com/blockchain-certificates/cert-verifier-js
https://github.com/digitalbazaar/jsonld.js
https://github.com/jsonld-java/jsonld-java

From memory, what I can remember is that cert verifier was somehow manually injecting JSONLD context when parsing the JSON LD hence if you used a parser without doing that manual addition, you won't be able to parse the JSON LD because of those missing fields and validate a certificate.

Though take this response with a pinch of salt , lots of time has passed!

lemoustachiste commented:
i am going to close this PR. It does not seem to be a longstanding issue and this might break existing certs if we merge so I'm going to stay on the safe side of things.

Plus we now prefer using v3 generally speaking
